ibis/backends/dask/execution/join.py

```diff
@@ -0,0 +1,109 @@
+import operator
+
+import dask.dataframe as dd
+from pandas import Timedelta
+
+import ibis.expr.operations as ops
+import ibis.util
+from ibis.backends.pandas.execution.join import (
+    _compute_join_column,
+    _extract_predicate_names,
+    _validate_columns,
+    execute_node,
+)
+
+from ..execution import constants
+
+
+@execute_node.register(
+    ops.AsOfJoin, dd.DataFrame, dd.DataFrame, (Timedelta, type(None))
+)
+def execute_asof_join(op, left, right, tolerance, **kwargs):
+    overlapping_columns = frozenset(left.columns) & frozenset(right.columns)
+    left_on, right_on = _extract_predicate_names(op.predicates)
+    left_by, right_by = _extract_predicate_names(op.by)
+    _validate_columns(
+        overlapping_columns, left_on, right_on, left_by, right_by
+    )
+
+    return dd.merge_asof(
+        left=left,
+        right=right,
+        left_on=left_on,
+        right_on=right_on,
+        left_by=left_by or None,
+        right_by=right_by or None,
+        tolerance=tolerance,
+    )
+
+
+@execute_node.register(ops.CrossJoin, dd.DataFrame, dd.DataFrame)
+def execute_cross_join(op, left, right, **kwargs):
+    """Execute a cross join in dask.
+
+    Notes
+    -----
+    We create a dummy column of all :data:`True` instances and use that as the
+    join key. This results in the desired Cartesian product behavior guaranteed
+    by cross join.
+
+    """
+    # generate a unique name for the temporary join key
+    key = "cross_join_{}".format(ibis.util.guid())
+    join_key = {key: True}
+    new_left = left.assign(**join_key)
+    new_right = right.assign(**join_key)
+
+    # inner/outer doesn't matter because every row matches every other row
+    result = dd.merge(
+        new_left,
+        new_right,
+        how='inner',
+        on=key,
+        suffixes=constants.JOIN_SUFFIXES,
+    )
+
+    # remove the generated key
+    del result[key]
+
+    return result
+
+
+# TODO - execute_materialized_join - #2553
+@execute_node.register(ops.Join, dd.DataFrame, dd.DataFrame)
+def execute_materialized_join(op, left, right, **kwargs):
```

Comment:
What is materialized join?

Reply:
I believe this is the join handler for joins on columns we need to materialize, i.e the following from `test_join_with_non_trivial_key`:

```python
    join = left.join(right, right.key.length() == left.key.length(), how=how)
    expr = join[left.key, left.value, right.other_value]
    result = expr.execute()

    expected = (
        dd.merge(
            df1.assign(key_len=df1.key.str.len()),
            df2.assign(key_len=df2.key.str.len()),
            on='key_len',
            how=how,
        )
        .drop(['key_len', 'key_y', 'key2', 'key3'], axis=1)
        .rename(columns={'key_x': 'key'})
    )
```

(code is basically https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/execution/join.py#L58),